### ISSUE TITLE
date_diff function returns incorrect results for DateTime64 arguments

### ISSUE THREAD

**Issue by rvasin** [ClickHouse/ClickHouse#41978]
**Describe what's wrong**

When both arguments for date_diff function are DateTime64 and start date is before 1970 and end date is after 1970 the function returns negative values.

**Does it reproduce on recent release?**

Yes.

**How to reproduce**

`select date_diff('day', toDateTime64('1960-01-01 10:00:00.000', 3), toDateTime64('1980-01-01 10:00:00.000', 3))`

Returns: -58231

**Expected behavior**

It should return 7304

**Additional context**

The following two queries (when both arguments before or both after 1970) work fine:

`select date_diff('day', toDateTime64('1970-01-01 10:00:00.000', 3), toDateTime64('1980-01-01 10:00:00.000', 3))`

Returns: 3652 

`select date_diff('day', toDateTime64('1950-01-01 10:00:00.000', 3), toDateTime64('1960-01-01 10:00:00.000', 3))`

Returns: 3652

**Comment by canhld94** [comment]
Related #22927
Can be fixed by let functions `toRelativeXNum` to return negative number if input is before Linux epoch time. I think currently they're returning overflowed numbers. @alexey-milovidov 

```sql
SELECT toRelativeDayNum(toDateTime('1969-12-31 23:59:59', 3))

┌─toRelativeDayNum(toDateTime('1969-12-31 23:59:59', 3))─┐
│                                                  65535 │
└────────────────────────────────────────────────────────┘

```

**Comment by rvasin** [comment]
@canhld94 yes, you are absolutely right. I verified it's related to toRelativeDayNum it returns UInt16. So for dates before 1970 it causes overflow.
I continued the discussion of the issue in #41977

**Comment by rvasin** [comment]
Closed by #42308
